go-controller/pkg/kube/annotator.go

```diff
@@ -0,0 +1,102 @@
+package kube
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	kapi "k8s.io/api/core/v1"
+)
+
+// Annotator represents the exported methods for handling node annotations
+type Annotator interface {
+	Set(key string, value interface{}) error
+	SetWithFailureHandler(key string, value interface{}, failFn FailureHandlerFn) error
+	Del(key string)
+	Run() error
+}
+
+// FailureHandlerFn is a function called when adding an annotation fails
+type FailureHandlerFn func(node *kapi.Node, key string, val interface{})
+
+type action struct {
+	key    string
+	val    interface{}
+	failFn FailureHandlerFn
+}
+
+type nodeAnnotator struct {
+	kube Interface
+	node *kapi.Node
+
+	changes map[string]*action
+}
+
+// NewNodeAnnotator returns a new annotator for Node objects
+func NewNodeAnnotator(kube Interface, node *kapi.Node) Annotator {
+	return &nodeAnnotator{
+		kube:    kube,
+		node:    node,
+		changes: make(map[string]*action),
+	}
+}
+
+func (na *nodeAnnotator) Set(key string, val interface{}) error {
+	return na.SetWithFailureHandler(key, val, nil)
+}
+
+func (na *nodeAnnotator) SetWithFailureHandler(key string, val interface{}, failFn FailureHandlerFn) error {
+	if val != nil && reflect.TypeOf(val).Kind() != reflect.String {
+		// Marshal to a string value which is what annotations require
+		bytes, err := json.Marshal(val)
+		if err != nil {
+			return fmt.Errorf("failed to marshal %q value %v to string: %v", key, val, err)
+		}
+		val = string(bytes)
+	}
+
+	na.changes[key] = &action{
+		key:    key,
+		val:    val,
+		failFn: failFn,
+	}
+
+	return nil
+}
+
+func (na *nodeAnnotator) Del(key string) {
+	na.changes[key] = &action{key: key}
+}
+
+func (na *nodeAnnotator) Run() error {
+	changes := make(map[string]interface{})
+	for k, act := range na.changes {
+		// Ignore annotations that already exist with the same value
+		if existing, ok := na.node.Annotations[k]; existing != act.val || !ok {
+			changes[k] = act.val
+		}
+	}
+	if len(changes) == 0 {
+		return nil
+	}
+
+	if err := na.kube.SetAnnotationsOnNode(na.node, changes); err != nil {
+		// Let failure handlers clean up
+		for _, act := range na.changes {
+			if act.failFn != nil {
+				act.failFn(na.node, act.key, act.val)
+			}
+		}
+		return err
+	}
+
+	for k, act := range na.changes {
+		if act.val == nil {
+			delete(na.node.Annotations, k)
+		} else {
+			na.node.Annotations[k] = act.val.(string)
```

Comment:
Since we're doing a merge, it's possible that another client simultaneously made other changes to the annotations that didn't conflict with ours. The Patch call returns an updated copy of the object. We should either use that, or else consider `na.node` to be out of date after the `Run` call.

Reply:
@danwinship turns out nothing actually needs the returned object at this time, so I removed the code that updated the old object, and skipped returning the new one. We can do that if we come across that case.